### DISCUSSION TITLE
Unable to expand product with subcategories?

### DISCUSSION THREAD

**Discussion by CODE14-algemeen** [swellstores/community#107] [Help] [answered]
If I fetch a product like so; 
```
const product = await swell.products.get(lastRoute, {
      expand: ['categories']`
});
```

I don't get the subcategories it's linked to, only the parent (head) category.

Also, I can't expand by adding
`expand: ['category']`

**Comment by ericingram** [comment]
Expanding `categories` should give you a collection of specific categories linked to a product. It doesn't return a full hierarchy, but it's possible to get that as well with more code. The `categories` field is defined like this for example:

```
    "categories": {
      "type": "link",
      "model": "categories:products",
      "cascade": true,
      "params": {
        "product_id": "id"
      }
    },
```

So it's a simple link to the join collection `categories:products`, where `product_id` matches the product.

**Reply by CODE14-algemeen** [reply]
But I can't fetch all the full hierarchy of categories straight from the product get request?

So I would have to fetch all categories and check which ones contain the product id?

**Reply by CODE14-algemeen** [reply]
Also, as stated in the docs, I don't have a 'category_index' property on my product model, how come?

**Reply by ericingram** [reply]
That's because `category_index` is a private field on the frontend API by default (using Swell-js).

Currently, there's no simple way to get the full hierarchy of a specific product in one query. There is a way to get all the top-level categories, though:

```
const product = await swell.products.get(productId, {
  $categories: true
});
```

But that doesn't return the full hierarchy either yet.

**Reply by CODE14-algemeen** [reply]
Is this a planned feature? I think more users would want this to be able to create things like breadcrumbs on product detail pages.

**Reply by CODE14-algemeen** [reply]
How would I be able to get the category and subcategory on a product detail page using the front-end API?

**Reply by ericingram** [reply]
You need to write a function to recursively fetch categories based on the top-level ID.

I do agree that we should implement a more complete "full breadcrumb" operation, to simplify things, but we're pretty busy with other stuff right now so it's not something you should wait around for.

**Comment by logeshswell** [comment]
@CODE14-algemeen cc: @ericingram Yes, you could do it by expanding as follows:
`expand: ['categories.parent.children']`

**Reply by CODE14-algemeen** [reply]
This doesn't work..

```
    product.value = await swell.products.get(lastRoute, {
        expand: ['categories.parent.children']
    });
```

Crashes the page.

**Reply by CODE14-algemeen** [reply]
@logeshswell Hi Logesh, this doesn't work on my front-end application.

How do I make this work? On either the swell-js package and the nuxt-swell package I don't get a response.

**Reply by logeshswell** [reply]
The above solution would work for swell-node (or backend API) package.
For swell-js, you need to open up the permissions to fetch these fields as follows:
https://developers.swell.is/guides/opening-content-permissions

**Reply by CODE14-algemeen** [reply]
I'm trying to figure out how to do this, i've tried multiple queries now but I can't seem to get it right?

Is there a possibility you can give me the right query to execute to be able to get the parent children?

**Reply by logeshswell** [reply, marked as answer]
Please run the below query for opening up content permissions:
```json
{
  "permissions": {
    "products": {
      "fields": ["breadcrumbs"],
      "expands": {
        "breadcrumbs": {
          "url": "/categories:products",
          "params": {
            "product_id": "id"
          },
          "data": {
            "expand": ["parent.children"]
          }
        }
      }
    }
  }
}
```

<img width="516" alt="Screenshot 2023-06-15 at 2 41 38 AM" src="https://github.com/swellstores/community/assets/100766131/2cdc5d4f-3166-4577-977f-770ab754584b">

Please execute the below code snippet:
```js
import swell from 'swell-js'

swell.init('store-id', 'public-key');

const products = await swell.products.get(product_id, {
  expand: ['breadcrumbs']
});
```

**Comment by jezzzm** [comment]
hey @CODE14-algemeen , because of this, I've ended up instead just fetching once and aggressively caching the full list of of categories instead of per-product.

you can use some of my helpers to manipulate the response from the /categories endpoint

https://gist.github.com/jezzzm/cf712330aecb952cffc56487876984a5

**Comment by ericingram** [comment]
As a side note, we're thinking about how to improve the frontend API experience around permissions. Ideas welcome!

**Reply by CODE14-algemeen** [reply]
I like the possibility to have modular model responses by using the 'expand' option, but I don't want to have to activate it through the developer console.

**Comment by christiancode14** [comment]
Any update on this? @ericingram

**Reply by ericingram** [reply]
No specific update. It will take time for us to plan out the improvements I mentioned, but the discussion is happening. In the meantime, there are a lot of good workarounds to solve querying issues that we can expand on as needed here.
